File: api/utils/password_validation.py
"""Password validation utilities."""

from __future__ import annotations

import re
from typing import Dict, List


def validate_password_strength(password: str) -> Dict[str, any]:
    """
    Validate password strength and return detailed feedback.

    Args:
        password: The password to validate

    Returns:
        Dict with:
            - valid: bool - Whether password meets all requirements
            - errors: List[str] - List of validation errors
            - score: int - Password strength score (0-5)
    """
    errors: List[str] = []
    score = 0

    # Check minimum length
    if len(password) < 8:
        errors.append("Password must be at least 8 characters long")
    else:
        score += 1

    # Check for uppercase letter
    if not re.search(r'[A-Z]', password):
        errors.append("Password must contain at least one uppercase letter")
    else:
        score += 1

    # Check for lowercase letter
    if not re.search(r'[a-z]', password):
        errors.append("Password must contain at least one lowercase letter")
    else:
        score += 1

    # Check for number
    if not re.search(r'[0-9]', password):
        errors.append("Password must contain at least one number")
    else:
        score += 1

    # Check for special character (optional but recommended)
    if re.search(r'[^A-Za-z0-9]', password):
        score += 1

    # Bonus for length
    if len(password) >= 12:
        score = min(score + 1, 5)

    return {
        "valid": len(errors) == 0,
        "errors": errors,
        "score": score,
        "strength": get_strength_label(score)
    }


def get_strength_label(score: int) -> str:
    """Get a human-readable strength label for a password score."""
    if score <= 1:
        return "Very Weak"
    elif score == 2:
        return "Weak"
    elif score == 3:
        return "Fair"
    elif score == 4:
        return "Strong"
    else:
        return "Very Strong"


def validate_username(username: str) -> Dict[str, any]:
    """
    Validate username format.

    Args:
        username: The username to validate

    Returns:
        Dict with valid (bool) and errors (List[str])
    """
    errors: List[str] = []

    # Check length
    if len(username) < 3:
        errors.append("Username must be at least 3 characters long")
    elif len(username) > 30:
        errors.append("Username must be 30 characters or less")

    # Check format (alphanumeric and underscores only)
    if not re.match(r'^[a-zA-Z0-9_]+$', username):
        errors.append("Username can only contain letters, numbers, and underscores")

    # Check doesn't start with number
    if username and username[0].isdigit():
        errors.append("Username cannot start with a number")

    return {
        "valid": len(errors) == 0,
        "errors": errors
    }


__all__ = ["validate_password_strength", "validate_username", "get_strength_label"]

File: src/utils/passwordValidation.ts
/**
 * Password validation utilities for frontend
 */

export interface PasswordValidation {
  valid: boolean;
  errors: string[];
  score: number;
  strength: string;
}

export interface UsernameValidation {
  valid: boolean;
  errors: string[];
}

export function validatePasswordStrength(password: string): PasswordValidation {
  const errors: string[] = [];
  let score = 0;

  // Check minimum length
  if (password.length < 8) {
    errors.push('Password must be at least 8 characters long');
  } else {
    score += 1;
  }

  // Check for uppercase letter
  if (!/[A-Z]/.test(password)) {
    errors.push('Password must contain at least one uppercase letter');
  } else {
    score += 1;
  }

  // Check for lowercase letter
  if (!/[a-z]/.test(password)) {
    errors.push('Password must contain at least one lowercase letter');
  } else {
    score += 1;
  }

  // Check for number
  if (!/[0-9]/.test(password)) {
    errors.push('Password must contain at least one number');
  } else {
    score += 1;
  }

  // Check for special character (optional but recommended)
  if (/[^A-Za-z0-9]/.test(password)) {
    score += 1;
  }

  // Bonus for length
  if (password.length >= 12) {
    score = Math.min(score + 1, 5);
  }

  return {
    valid: errors.length === 0,
    errors,
    score,
    strength: getStrengthLabel(score),
  };
}

export function getStrengthLabel(score: number): string {
  if (score <= 1) return 'Very Weak';
  if (score === 2) return 'Weak';
  if (score === 3) return 'Fair';
  if (score === 4) return 'Strong';
  return 'Very Strong';
}

export function getStrengthColor(score: number): string {
  if (score <= 1) return '#d32f2f';
  if (score === 2) return '#f57c00';
  if (score === 3) return '#fbc02d';
  if (score === 4) return '#689f38';
  return '#388e3c';
}

export function validateUsername(username: string): UsernameValidation {
  const errors: string[] = [];

  // Check length
  if (username.length < 3) {
    errors.push('Username must be at least 3 characters long');
  } else if (username.length > 30) {
    errors.push('Username must be 30 characters or less');
  }

  // Check format (alphanumeric and underscores only)
  if (!/^[a-zA-Z0-9_]+$/.test(username)) {
    errors.push('Username can only contain letters, numbers, and underscores');
  }

  // Check doesn't start with number
  if (username && /^[0-9]/.test(username)) {
    errors.push('Username cannot start with a number');
  }

  return {
    valid: errors.length === 0,
    errors,
  };
}
